odoo/service/server.py

```diff
@@ -833,17 +835,19 @@ def sleep(self):
         try:
             # map of fd -> worker
             fds = {w.watchdog_pipe[0]: w for w in self.workers.values()}
-            fd_in = list(fds) + [self.pipe[0]]
             # check for ping or internal wakeups
-            ready = select.select(fd_in, [], [], self.beat)
-            # update worker watchdogs
-            for fd in ready[0]:
-                if fd in fds:
-                    fds[fd].watchdog_time = time.time()
-                empty_pipe(fd)
-        except select.error as e:
-            if e.args[0] not in [errno.EINTR]:
-                raise
+            with select() as sel:
+                for fd in fds:
+                    sel.register(fd, selectors.EVENT_READ)
+                sel.register(self.pipe[0], selectors.EVENT_READ)
+                events = sel.select(timeout=self.beat)
```

Comment:
Here you create a new `select()` object and you re-register all the file descriptions every time we `sleep`. It would be better to create the select object once and to pro-actively add/remove file-description as they spawn and die.